# INTRODUCTION

# BUSINESS uNDERSTANDING

# PROBLEM STATEMENT

# DATA uNDERSRTANDING
This dataset represents the median monthly home sales prices for 265 zip codes from April 1996 to April 2018 according to Zillow statistics.

A distinct zip code is represented by each row. Each record includes area details and monthly median sales prices for homes.

There are 14,723 rows and 272 variables:

* `RegionID`: Unique index, 58196 through 753844
* `RegionName`: Unique Zip Code, 1001 through 99901
* `City`: City in which the zip code is located
* `State`: State in which the zip code is located
* `Metro`: Metropolitan Area in which the zip code is located
* `CountyName`: County in which the zip code is located
* `SizeRank`: Numerical rank of size of zip code, ranked 1 through 14723
* `1996-04 through 2018-04`: refers to the median housing sales values for April 1996 through April 2018, that is 265 data points of monthly data for each zip code

# IMPORTING LIBRARIES

In [30]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import itertools
import time
# from user_functions import *
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

# READING THE DATASETS

In [31]:

pd.set_option("display.max_columns", None)

zillow_df = pd.read_csv('zillow_data.csv')
zillow_df.head()

RegionID  RegionName      City State              Metro CountyName  \
0     84654       60657   Chicago    IL            Chicago       Cook   
1     90668       75070  McKinney    TX  Dallas-Fort Worth     Collin   
2     91982       77494      Katy    TX            Houston     Harris   
3     84616       60614   Chicago    IL            Chicago       Cook   
4     93144       79936   El Paso    TX            El Paso    El Paso   

   SizeRank   1996-04   1996-05   1996-06   1996-07   1996-08   1996-09  \
0         1  334200.0  335400.0  336500.0  337600.0  338500.0  339500.0   
1         2  235700.0  236900.0  236700.0  235400.0  233300.0  230600.0   
2         3  210400.0  212200.0  212200.0  210700.0  208300.0  205500.0   
3         4  498100.0  500900.0  503100.0  504600.0  505500.0  505700.0   
4         5   77300.0   77300.0   77300.0   77300.0   77400.0   77500.0   

    1996-10   1996-11   1996-12   1997-01   1997-02   1997-03   1997-04  \
0  340400.0  341300.0  342600.0  344400.0  345700.0  346700.0  347800.0   
1  227300.0  223400.0  219600.0  215800.0  211100.0  205700.0  200900.0   
2  202500.0  199800.0  198300.0  197300.0  195400.0  193000.0  191800.0   
3  505300.0  504200.0  503600.0  503400.0  502200.0  500000.0  497900.0   
4   77600.0   77700.0   77700.0   77800.0   77900.0   77900.0   77800.0   

    1997-05   1997-06   1997-07   1997-08   1997-09   1997-10   1997-11  \
0  349000.0  350400.0  352000.0  353900.0  356200.0  358800.0  361800.0   
1  196800.0  193600.0  191400.0  190400.0  190800.0  192700.0  196000.0   
2  191800.0  193000.0  195200.0  198400.0  202800.0  208000.0  213800.0   
3  496300.0  495200.0  494700.0  494900.0  496200.0  498600.0  502000.0   
4   77800.0   77800.0   77800.0   77800.0   77900.0   78100.0   78200.0   

    1997-12   1998-01   1998-02   1998-03   1998-04   1998-05   1998-06  \
0  365700.0  370200.0  374700.0  378900.0  383500.0  388300.0  393300.0   
1  201300.0  207400.0  212200.0  214600.0  215100.0  213400.0  210200.0   
2  220700.0  227500.0  231800.0  233400.0  233900.0  233500.0  233300.0   
3  507600.0  514900.0  522200.0  529500.0  537900.0  546900.0  556400.0   
4   78400.0   78600.0   78800.0   79000.0   79100.0   79200.0   79300.0   

    1998-07   1998-08   1998-09   1998-10   1998-11   1998-12   1999-01  \
0  398500.0  403800.0  409100.0  414600.0  420100.0  426200.0  432600.0   
1  206100.0  202100.0  198800.0  196100.0  194100.0  193400.0  193400.0   
2  234300.0  237400.0  242800.0  250200.0  258600.0  268000.0  277000.0   
3  566100.0  575600.0  584800.0  593500.0  601600.0  610100.0  618600.0   
4   79300.0   79300.0   79400.0   79500.0   79500.0   79600.0   79700.0   

    1999-02   1999-03   1999-04   1999-05   1999-06   1999-07   1999-08  \
0  438600.0  444200.0  450000.0  455900.0  462100.0  468500.0  475300.0   
1  193100.0  192700.0  193000.0  193700.0  194800.0  196100.0  197800.0   
2  283600.0  288500.0  293900.0  299200.0  304300.0  308600.0  311400.0   
3  625600.0  631100.0  636600.0  642100.0  647600.0  653300.0  659300.0   
4   79900.0   80100.0   80300.0   80600.0   80900.0   81200.0   81400.0   

    1999-09   1999-10   1999-11   1999-12   2000-01   2000-02   2000-03  \
0  482500.0  490200.0  498200.0  507200.0  516800.0  526300.0  535300.0   
1  199700.0  201900.0  204500.0  207800.0  211500.0  214900.0  217800.0   
2  312300.0  311900.0  311100.0  311700.0  313500.0  315000.0  316700.0   
3  665800.0  672900.0  680500.0  689600.0  699700.0  709300.0  718300.0   
4   81700.0   82100.0   82400.0   82600.0   82800.0   82900.0   83000.0   

    2000-04   2000-05   2000-06   2000-07   2000-08   2000-09   2000-10  \
0  544500.0  553500.0  562400.0  571200.0  579800.0  588100.0  596300.0   
1  221100.0  224100.0  226700.0  228200.0  228500.0  227200.0  224900.0   
2  319800.0  323700.0  327500.0  329900.0  329800.0  326400.0  320100.0   
3  727600.0  737100.0  746600.0  756200.0  765800.0  775100.0  784400.0   
4   83000.0   82900.0   82800.0 

In [32]:
zillow_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14723 entries, 0 to 14722
Columns: 272 entries, RegionID to 2018-04
dtypes: float64(219), int64(49), object(4)
memory usage: 30.6+ MB


In [33]:
# Check for missing values
zillow_df.isna().sum()

RegionID         0
RegionName       0
City             0
State            0
Metro         1043
              ... 
2017-12          0
2018-01          0
2018-02          0
2018-03          0
2018-04          0
Length: 272, dtype: int64

We can see that the `Metro` column has many missing values. Further investigation may be necessary to learn to deal with this column.

# DATA CLEANING
The initial step is to investigate each column. 
1. Analyzing the `RegionID` column which is the unique indices.

In [34]:
print(f"Count of unique entries", zillow_df.RegionID.value_counts())
print(f"Number of Unique values", zillow_df.RegionID.nunique())

Count of unique entries 73724     1
70551     1
99221     1
76688     1
82829     1
         ..
71176     1
91654     1
65029     1
100380    1
98304     1
Name: RegionID, Length: 14723, dtype: int64
Number of Unique values 14723


We can see that each entry is unique values maybe describing the different zipcodes of the properties. It would thus be necessary to change this value to a string data type.

In [35]:
zillow_df.RegionID = zillow_df.RegionID.astype('string')

2. Analyzing `Region Name` column: 
Unique Zip Code, 1001 through 99901 it it would be best to cast to a string. This is an important aspect of the modeling of this project

In [36]:
zillow_df.RegionName.value_counts()

55324    1
74561    1
73538    1
31546    1
82070    1
        ..
75182    1
55343    1
1450     1
73129    1
65536    1
Name: RegionName, Length: 14723, dtype: int64

In [37]:
zillow_df.RegionName = zillow_df.RegionName.astype('string')

In [51]:
# change the region name to the appropriate zipcodes between 1001 to 99901
for i in range(len(zillow_df)):
    zillow_df.RegionName[i] = zillow_df.RegionName[i].rjust(5, '0')

zillow_df[['RegionName', 'State']].sort_values(by=['RegionName'])
zillow_df.RegionName = zillow_df.RegionName.astype('int')



We can see that the `RegionName` column has been change to the unique codes of the different regions/ ststes between 01001 to 99901

3. Analyzing `State` column

In [52]:
print(f"Count of unique entries", zillow_df.State.value_counts())
print(" ")
print(f"Number of Unique values", zillow_df.State.nunique())

Count of unique entries CA    1224
NY    1015
TX     989
PA     831
FL     785
OH     588
IL     547
NJ     502
MI     499
NC     428
IN     428
MA     417
TN     404
VA     401
MN     375
GA     345
WA     341
WI     332
MO     319
MD     317
CO     249
KS     241
AZ     230
OR     224
OK     221
SC     206
NH     199
LA     193
AL     183
IA     158
MS     153
KY     139
CT     124
UT     121
ID     110
AR     105
NV     103
ME      86
NE      83
WV      72
MT      71
HI      62
NM      60
RI      59
DE      41
WY      31
ND      31
AK      28
SD      19
DC      18
VT      16
Name: State, dtype: int64
 
Number of Unique values 51


4. Analyzing the `Metro` column which is the Metropolitan Area in which the zip code is located. 
Initial data exploration showed the presence of 1043 NaN values. It msybe be best to deal with these. Most probable appraoch would be to fill these with None meaning that there is no Metropolitan Area in the given zipcode. Perharps the presence of a Metrpolitan Area has an infleunce on the value of the 

In [53]:
# Deal with missing values
zillow_df.Metro.fillna('None', inplace=True)

In [54]:
print(f"Count of unique entries", zillow_df.Metro.value_counts())

print(f"Number of Unique values", zillow_df.Metro.nunique())


Count of unique entries None                              1043
New York                           779
Los Angeles-Long Beach-Anaheim     347
Chicago                            325
Philadelphia                       281
                                  ... 
Milledgeville                        1
New Ulm                              1
Sweetwater                           1
Wahpeton                             1
North Platte                         1
Name: Metro, Length: 702, dtype: int64
Number of Unique values 702


5. Analyzing the sales columns in the columns `1996-04 through 2018-04`.  which refers to the median housing sales values for April 1996 through April 2018, that is 265 data points of monthly data for each zip code.
This column is important in choosing the ROI yield.

The presence of NaN values in some of columns can make the model assume that the there were no sales in the zipcodes at those period. Which can be an incorrect assumption.

The period from 1996-04 to 2018-04 is a 22 year period maybe selecting different periods i.e 4 year median sales value, average median sales, 1 year median sales values.

In [55]:
# Create a column calcualting the ROI for a 5 year period.

zillow_df['1_yr_ROI'] = (zillow_df['2018-04'] - zillow_df['2017-04'])/(zillow_df['2017-04'])
zillow_df

RegionID  RegionName                 City State              Metro  \
0        84654       60657              Chicago    IL            Chicago   
1        90668       75070             McKinney    TX  Dallas-Fort Worth   
2        91982       77494                 Katy    TX            Houston   
3        84616       60614              Chicago    IL            Chicago   
4        93144       79936              El Paso    TX            El Paso   
...        ...         ...                  ...   ...                ...   
14718    58333        1338             Ashfield    MA    Greenfield Town   
14719    59107        3293            Woodstock    NH          Claremont   
14720    75672       40404                Berea    KY           Richmond   
14721    93733       81225  Mount Crested Butte    CO               None   
14722    95851       89155             Mesquite    NV          Las Vegas   

      CountyName  SizeRank   1996-04   1996-05   1996-06   1996-07   1996-08  \
0           Cook         1  334200.0  335400.0  336500.0  337600.0  338500.0   
1         Collin         2  235700.0  236900.0  236700.0  235400.0  233300.0   
2         Harris         3  210400.0  212200.0  212200.0  210700.0  208300.0   
3           Cook         4  498100.0  500900.0  503100.0  504600.0  505500.0   
4        El Paso         5   77300.0   77300.0   77300.0   77300.0   77400.0   
...          ...       ...       ...       ...       ...       ...       ...   
14718   Franklin     14719   94600.0   94300.0   94000.0   93700.0   93400.0   
14719    Grafton     14720   92700.0   92500.0   92400.0   92200.0   92100.0   
14720    Madison     14721   57100.0   57300.0   57500.0   57700.0   58000.0   
14721   Gunnison     14722  191100.0  192400.0  193700.0  195000.0  196300.0   
14722      Clark     14723  176400.0  176300.0  176100.0  176000.0  175900.0   

        1996-09   1996-10   1996-11   1996-12   1997-01   1997-02   1997-03  \
0      339500.0  340400.0  341300.0  342600.0  344400.0  345700.0  346700.0   
1      230600.0  227300.0  223400.0  219600.0  215800.0  211100.0  205700.0   
2      205500.0  202500.0  199800.0  198300.0  197300.0  195400.0  193000.0   
3      505700.0  505300.0  504200.0  503600.0  503400.0  502200.0  500000.0   
4       77500.0   77600.0   77700.0   77700.0   77800.0   77900.0   77900.0   
...         ...       ...       ...       ...       ...       ...       ...   
14718   93200.0   93000.0   92900.0   92700.0   92600.0   92600.0   92600.0   
14719   91900.0   91700.0   91300.0   90900.0   90500.0   90100.0   89800.0   
14720   58200.0   58400.0   58700.0   59100.0   59500.0   59900.0   60300.0   
14721  197700.0  199100.0  200700.0  202600.0  204900.0  207100.0  209100.0   
14722  175800.0  175800.0  176000.0  176200.0  176500.0  176700.0  176800.0   

        1997-04   1997-05   1997-06   1997-07   1997-08   1997-09   1997-10  \
0      347800.0  349000.0  350400.0  352000.0  353900.0  356200.0  358800.0   
1      200900.0  196800.0  193600.0  191400.0  190400.0  190800.0  192700.0   
2      191800.0  191800.0  193000.0  195200.0  198400.0  202800.0  208000.0   
3      497900.0  496300.0  495200.0  494700.0  494900.0  496200.0  498600.0   
4       77800.0   77800.0   77800.0   77800.0   77800.0   77900.0   78100.0   
...         ...       ...       ...       ...       ...       ...       ...   
14718   92800.0   93100.0   93500.0   94000.0   94500.0   95100.0   95800.0   
14719   89600.0   89400.0   89300.0   89200.0   89100.0   88900.0   88700.0   
14720   60800.0   61300.0   62000.0   62600.0   63200.0   63800.0   64500.0   
14721  211200.0  213400.0  215800.0  218300.0  221000.0  223900.0  226900.0   
14722  176800.0  176900.0  177000.0  177100.0  177200.0  177400.0  177600.0   

        1997-11   1997-12   1998-01   1998-02   1998-03   1998-04   1998-05  \
0      361800.0  365700.0  370200.0  374700.0  378900.0  383500.0  388300.0   
1      196000.0  201300.0  207400.0  212200.0  214600.0  215100.0  213400.0 

In [56]:
# Create a column calcualting the ROI for a 5 year period.

zillow_df['5_yr_ROI'] = (zillow_df['2018-04'] - zillow_df['2013-04'])/(zillow_df['2014-07'])


In [57]:
# Create a column calcualting the ROI for a 10 year period.

zillow_df['10_yr_ROI'] = (zillow_df['2018-04'] - zillow_df['2008-04'])/(zillow_df['2008-04'])
zillow_df

RegionID  RegionName                 City State              Metro  \
0        84654       60657              Chicago    IL            Chicago   
1        90668       75070             McKinney    TX  Dallas-Fort Worth   
2        91982       77494                 Katy    TX            Houston   
3        84616       60614              Chicago    IL            Chicago   
4        93144       79936              El Paso    TX            El Paso   
...        ...         ...                  ...   ...                ...   
14718    58333        1338             Ashfield    MA    Greenfield Town   
14719    59107        3293            Woodstock    NH          Claremont   
14720    75672       40404                Berea    KY           Richmond   
14721    93733       81225  Mount Crested Butte    CO               None   
14722    95851       89155             Mesquite    NV          Las Vegas   

      CountyName  SizeRank   1996-04   1996-05   1996-06   1996-07   1996-08  \
0           Cook         1  334200.0  335400.0  336500.0  337600.0  338500.0   
1         Collin         2  235700.0  236900.0  236700.0  235400.0  233300.0   
2         Harris         3  210400.0  212200.0  212200.0  210700.0  208300.0   
3           Cook         4  498100.0  500900.0  503100.0  504600.0  505500.0   
4        El Paso         5   77300.0   77300.0   77300.0   77300.0   77400.0   
...          ...       ...       ...       ...       ...       ...       ...   
14718   Franklin     14719   94600.0   94300.0   94000.0   93700.0   93400.0   
14719    Grafton     14720   92700.0   92500.0   92400.0   92200.0   92100.0   
14720    Madison     14721   57100.0   57300.0   57500.0   57700.0   58000.0   
14721   Gunnison     14722  191100.0  192400.0  193700.0  195000.0  196300.0   
14722      Clark     14723  176400.0  176300.0  176100.0  176000.0  175900.0   

        1996-09   1996-10   1996-11   1996-12   1997-01   1997-02   1997-03  \
0      339500.0  340400.0  341300.0  342600.0  344400.0  345700.0  346700.0   
1      230600.0  227300.0  223400.0  219600.0  215800.0  211100.0  205700.0   
2      205500.0  202500.0  199800.0  198300.0  197300.0  195400.0  193000.0   
3      505700.0  505300.0  504200.0  503600.0  503400.0  502200.0  500000.0   
4       77500.0   77600.0   77700.0   77700.0   77800.0   77900.0   77900.0   
...         ...       ...       ...       ...       ...       ...       ...   
14718   93200.0   93000.0   92900.0   92700.0   92600.0   92600.0   92600.0   
14719   91900.0   91700.0   91300.0   90900.0   90500.0   90100.0   89800.0   
14720   58200.0   58400.0   58700.0   59100.0   59500.0   59900.0   60300.0   
14721  197700.0  199100.0  200700.0  202600.0  204900.0  207100.0  209100.0   
14722  175800.0  175800.0  176000.0  176200.0  176500.0  176700.0  176800.0   

        1997-04   1997-05   1997-06   1997-07   1997-08   1997-09   1997-10  \
0      347800.0  349000.0  350400.0  352000.0  353900.0  356200.0  358800.0   
1      200900.0  196800.0  193600.0  191400.0  190400.0  190800.0  192700.0   
2      191800.0  191800.0  193000.0  195200.0  198400.0  202800.0  208000.0   
3      497900.0  496300.0  495200.0  494700.0  494900.0  496200.0  498600.0   
4       77800.0   77800.0   77800.0   77800.0   77800.0   77900.0   78100.0   
...         ...       ...       ...       ...       ...       ...       ...   
14718   92800.0   93100.0   93500.0   94000.0   94500.0   95100.0   95800.0   
14719   89600.0   89400.0   89300.0   89200.0   89100.0   88900.0   88700.0   
14720   60800.0   61300.0   62000.0   62600.0   63200.0   63800.0   64500.0   
14721  211200.0  213400.0  215800.0  218300.0  221000.0  223900.0  226900.0   
14722  176800.0  176900.0  177000.0  177100.0  177200.0  177400.0  177600.0   

        1997-11   1997-12   1998-01   1998-02   1998-03   1998-04   1998-05  \
0      361800.0  365700.0  370200.0  374700.0  378900.0  383500.0  388300.0   
1      196000.0  201300.0  207400.0  212200.0  214600.0  215100.0  213400.0 

VISUALIZING THE SPREAD OF THE ROI

TIME SERIES DATA PREPROCESSING.
It is necessary to change the 


# EXPLORATORY DATA ANALYSIS

font = {'family' : 'normal',
        'weight' : 'bold',
        'size'   : 22}

matplotlib.rc('font', **font)


RESHAPING TIME SERIES DATA
it would be necessary to chnge the dataframe from wide to long format

In [60]:
def melt_df(df):
    # Assuming 'Date' is the correct column name for date information
    melt = pd.melt(df, id_vars=[ 'RegionID', 'State', 'Metro', 'CountyName', '1_yr_ROI', '5_yr_ROI', '10_yr_ROI'], var_name='Date', value_name='value')
    
    # Ensure the 'Date' column is converted to datetime format
    melt['Date'] = pd.to_datetime(melt['Date'], errors='coerce',format='%Y-%m' )  # Use 'coerce' to handle invalid dates
    
    # Drop rows with missing date 
    melt = melt.dropna(subset=['value'])  # Ensure 'Date' and 'value' columns are present
    return melt



zillow_melted_df = melt_df(zillow_df)
zillow_melted_df

RegionID State              Metro CountyName  1_yr_ROI  5_yr_ROI  \
0          84654    IL            Chicago       Cook  0.041852  0.238799   
1          90668    TX  Dallas-Fort Worth     Collin  0.057162  0.465696   
2          91982    TX            Houston     Harris  0.030937  0.250430   
3          84616    IL            Chicago       Cook  0.019103  0.230110   
4          93144    TX            El Paso    El Paso  0.029661  0.070237   
...          ...   ...                ...        ...       ...       ...   
3945759    58333    MA    Greenfield Town   Franklin -0.010402  0.156118   
3945760    59107    NH          Claremont    Grafton  0.162120  0.300768   
3945761    75672    KY           Richmond    Madison  0.104305  0.333013   
3945762    93733    CO               None   Gunnison  0.121350  0.247446   
3945763    95851    NV          Las Vegas      Clark  0.100092  0.457277   

         10_yr_ROI       Date   value  
0         0.191307        NaT   60657  
1         0.523674        NaT   75070  
2         0.322244        NaT   77494  
3         0.142383        NaT   60614  
4        -0.053001        NaT   79936  
...            ...        ...     ...  
3945759   0.058140 2018-04-01  209300  
3945760  -0.005724 2018-04-01  225800  
3945761   0.292636 2018-04-01  133400  
3945762   0.040238 2018-04-01  664400  
3945763   0.048430 2018-04-01  357200  

[3788873 rows x 9 columns]

In [61]:
zillow_melted_df.describe()

1_yr_ROI      5_yr_ROI     10_yr_ROI
count  3.788873e+06  3.780652e+06  3.738050e+06
mean   7.118628e-02  3.084691e-01  1.462332e-01
std    5.755482e-02  1.977836e-01  2.150941e-01
min   -2.181352e-01 -3.064067e-01 -5.565217e-01
25%    3.628585e-02  1.663480e-01  6.027122e-03
50%    6.592040e-02  2.710901e-01  1.272523e-01
75%    9.965238e-02  4.176428e-01  2.540395e-01
max    5.080775e-01  1.347985e+00  1.664748e+00

# MODELING
We can first start to looking to the distribtuion of the ROI within the periods


Univariate analysis to see the distribution of house values

In [63]:
# distribtuion of 5 year ROI period
plt.figure(figsize=(15,4))
sns.histplot(zillow_melted_df['1_yr_ROI'],kde=True)
plt.title('Distribution 1yr ROI')
plt.ylabel('Density')
plt.xlabel('5yr ROI')
plt.show()

In [ ]:
# distribtuion of 5 year ROI period
plt.figure(figsize=(15,4))
sns.histplot(zillow_melted_df['5_yr_ROI'],kde=True)
plt.title('Distribution 5yr ROI')
plt.ylabel('Density')
plt.xlabel('5yr ROI')
plt.show()

In [ ]:
# distribtuion of 5 year ROI period
plt.figure(figsize=(15,4))
sns.histplot(zillow_melted_df['10_yr_ROI'],kde=True)
plt.title('Distribution 10yr ROI')
plt.ylabel('Density')
plt.xlabel('5yr ROI')
plt.show()

# ARIMA MODELING

# INTERPRETING RESULTS

# CONCLUSION